numpy/ma/core.py

```diff
@@ -4984,23 +4877,51 @@ def mean(self, axis=None, dtype=None, out=None):
 
         """
         if self._mask is nomask:
-            result = super(MaskedArray, self).mean(axis=axis, dtype=dtype)
-        else:
-            dsum = self.sum(axis=axis, dtype=dtype)
-            cnt = self.count(axis=axis)
-            if cnt.shape == () and (cnt == 0):
-                result = masked
+            ret = super(MaskedArray, self).mean(axis, dtype, out, **kwargs)
+            zerocounts = self.count(axis, **kwargs) == 0
+            if out is not None:
+                if isinstance(out, MaskedArray):
+                    if any(zerocounts):
+                        out.__setmask__(zerocounts)
+                    else:
+                        out.__setmask__(nomask)
+                return out
+            # mask in case of empty slices
+            if any(zerocounts):
+                ret = ret.view(MaskedArray)
+                ret.__setmask__(zerocounts)
+            return ret
+
+        # code largely copied from _methods.mean
+
+        rcount = self.count(axis=axis, **kwargs)
+
+        # Cast bool, unsigned int, and int to float64 by default
+        if dtype is None and issubclass(self.dtype.type,
+                                        (np.integer, np.bool_)):
+            dtype = np.dtype('f8')
+
+        ret = self.sum(axis=axis, dtype=dtype, **kwargs)
+        if ret is masked:
+            return _return_masked(out)
+        with np.errstate(divide='ignore', invalid='ignore'):
+            if isinstance(ret, np.ndarray):
+                ret = np.true_divide(
+                        ret, rcount, out=ret, casting='unsafe', subok=False)
```

Comment:
But any `ndarray` subclass will pass through the if-statement, so I think there should be a `subok=True`:
```
import numpy as np, astropy.units as u
q = np.arange(10.)*u.m
np.true_divide(q, 10, subok=False)
# array([ 0. ,  0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9])
np.true_divide(q, 10, subok=True)
# <Quantity [ 0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9] m>
```
Note that the default is `True`, so you can also just remove the `subok=False` part.